src/nvim/tui/tui.c

```diff
@@ -69,6 +70,9 @@ typedef struct {
   bool can_use_terminal_scroll;
   bool mouse_enabled;
   bool busy;
+  /* use int instead ? */
+  // cursorentry_T cursor_shapes[SHAPE_IDX_COUNT]; /**!< is it fine to reuse cursorentry_T here ? */
```

Comment:
cursorentry_T can be used (like HlAttrs below). This is TUI's internal implementation, it never sees RPC. Think of it as its own "external UI".